lib/stages/visSharedMemWriter.cpp

```diff
@@ -0,0 +1,389 @@
+    // takes the name of a shared memory address, opens a shared memory of the provided size, and maps the memory to a uint8_t* address pointer
+
+        uint8_t* addr;
+
+        int fd = shm_open(shm_name.c_str(), (O_CREAT | O_RDWR), (S_IRUSR | S_IWUSR));
+
+        if (fd == -1) {
+            throw std::runtime_error(
+                fmt::format(fmt("Cannot open shared memory named {:s}: {:s}"), shm_name, strerror(errno)));
+        }
+
+        // Resize object to hold buffer
+        if (ftruncate(fd, shm_size) == -1) {
+            throw std::runtime_error(
+                fmt::format(fmt("Failed to expand shared memory named {:s}: {:s}"), shm_name, strerror(errno)));
+        }
+        INFO("Resized to {} bytes\n", (long) shm_size);
+
+        addr = (uint8_t*) mmap(nullptr, shm_size, (PROT_READ | PROT_WRITE), MAP_SHARED, fd, 0);
+        if (addr == MAP_FAILED) {
+            throw std::runtime_error(
+                fmt::format(fmt("Failed to map shm {:s} to memory: {:s}."), shm_name, strerror(errno)));
+        }
+
+        // fd is no longer needed
+        if (close(fd) == -1) {
+            throw std::runtime_error(
+                fmt::format(fmt("Failed to close file descriptor for shm {:s}: {:s}."), shm_name, strerror(errno)));
+        }
+
+        return addr;
+}
+
+bool visSharedMemWriter::add_sample(const visFrameView& frame, time_ctype t, uint32_t freq_ind) {
+    // calculate the time index for time sample t, add the frame for time sample t at position frequency index
+    //
+    // curr_pos always points to the ring buffer index for the most recent time
+
+    if ( vis_time_ind_map.count(t) != 0) {
+        // if the time is already indexed, write to memory at that location
+        write_to_memory(frame, vis_time_ind_map.at(t), freq_ind);
+        return true;
+    }
+    else if (vis_time_ind_map.size() == 0) {
+        // the first sample added, so we do not increment by 1
+        vis_time_ind_map[t] = cur_pos;
+        write_to_memory(frame, vis_time_ind_map.at(t), freq_ind);
+        return true;
+    }
+    else if (vis_time_ind_map.size() < ntime) {
+        // if there is still empty space in the shared buffer
+        // add an additional time index
+        cur_pos++;
+        vis_time_ind_map[t] = cur_pos;
+        write_to_memory(frame, vis_time_ind_map.at(t), freq_ind);
+        return true;
+    }
+
+    // obtain the most recent and oldest time
+    time_ctype max_time = vis_time_ind_map.rbegin()->first;
+    time_ctype min_time = vis_time_ind_map.begin()->first;
+
+    if (t < min_time) {
+        // this data is older than anything else in the map, so we should
+        // just drop it
+        INFO("Dropping integration as buffer (FPGA count: {:d}) arrived too late (minimum in pool {:d})\n", t.fpga_count, min_time.fpga_count);
+        return false;
+    }
+
+    else if (t > max_time) {
+        // we need to drop the oldest time
+        cur_pos++;
+        reset_memory(cur_pos);
```

Comment:
why is this necessary?

Reply:
We want to wipe out the entire old time sample, before we write in a frame for the new time.

We have `time 0`. We received new `time 10`. We want to erase all frames at `time 0` before we replace it with a frame with `time 10` and `freq 200`.

Reply:
Does that make sense?

My write_to_memory function assumes writing to a specific `time_ind` and `freq_ind`. `reset_memory` wipes an entire `time_ind` in one go.

Reply:
ooooh, i see. Yes, make sense, thanks!